File: leetcode_oj/count_smaller.cpp
class Solution {
public:

void help(vector<int>& nums, vector<int>& vec, vector<int>& copy, int l, int r, int smaller[]) {

    if(l==r) return;

    int m = l+(r-l)/2;
    help(nums, vec, copy, l, m, smaller);
    help(nums, vec, copy, m+1, r, smaller);

    int i=l, j=m+1, k=l;
    while(i<=m || j<=r) {
        if( j==r+1 || (i<=m &&  nums[vec[i]] <= nums[vec[j]])) {
            copy[k++] = vec[i];
            smaller[vec[i]] += j-m-1;
            i++;
        }
        else copy[k++] = vec[j++];
    }

    for(int i=l; i<=r; i++) {
        vec[i] = copy[i];
    }
}

vector<int> countSmaller(vector<int>& nums) {
    vector<int> res;
    if(nums.empty()) return res;

    vector<int> vec, copy;
    for(int i=0; i<nums.size(); i++) {
        vec.push_back(i);
        copy.push_back(i);
    }


    int smaller[nums.size()];
    for(int i=0; i<nums.size(); i++) {
        smaller[i] = 0;
    }


    help(nums, vec, copy, 0, vec.size()-1, smaller);


    for(int i: smaller) {
        res.push_back(i);
    }

    return res;
}

};

File: leetcode_oj/count_smaller.py
class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        """
        Algorithm:
        Merge sort/Counting inversion: if left # jumps to right (or vice versa)
        Sort and then see how each number moved, for example 5 moved 
        two spots to the right so there were two smaller numbers on its right
        """
        self.smaller = [0] * len(nums)
        self.merge_sort(list(enumerate(nums)))
        return self.smaller

    def merge_sort(self, nums):
        half = len(nums)/2
        if half:
            l, r = self.merge_sort(nums[:half]),\
                    self.merge_sort(nums[half:])
            m = len(l); n = len(r)
            i = 0; j = 0
            while i < m or j < n:
                print "i: ", i, " j: ", j
                if j == n or i < m and l[i][1] <= r[j][1]:
                    print "l: ", l[i], 
                    print " r: ", r[j] if j < n else None
                    self.smaller[l[i][0]] += j
                    nums[i+j] = l[i]
                    i += 1
                else:
                    nums[i+j] = r[j]
                    j += 1
        return nums

s = Solution()
smaller = s.countSmaller([2,5,1,3,4])

